File: libs/std/modify.lua
require "SltRuntime"

function copy(obj, seen)
  -- Handle non-tables and previously-seen tables.
  if type(obj) ~= 'table' then return obj end
  if seen and seen[obj] then return seen[obj] end

  -- New table; mark it as seen and copy recursively.
  local s = seen or {}
  local res = {}
  s[obj] = res
  for k, v in pairs(obj) do res[copy(k, s)] = copy(v, s) end
  return setmetatable(res, getmetatable(obj))
end

baseModify = SltThunk.create(
  function() return 
    SltFunc.create(
      function(struct)
        return SltFunc.create(
          function(string)
            return SltFunc.create(
              function(func)
                local a = copy(struct())
                if not a.is_struct then SltError.create("first argument TypeError", "Must be a struct rather than being " .. a.type_, a) end
                local val = a.table[string().value]
                if val == nil then 
                  strStruct = tostring(struct())
                  print(strStruct)
                  SltError.create("first argument KeyError",  "Key ".. tostring(string()) .. " not found in " .. strStruct, a) 
                end
                a.table[string().value] = SltThunk.create(function() return func()(val) end)
                return a
              end
            )
          end
        )
      end     
    )
  end,
  Mutates
)

File: libs/std/modify.js
import { SltFunc, SltThunk, SltStruct, error } from "../../SltRuntime.ts"

const cloneFun = self => {
    var that = self;
    var temp = function temporary() { return that.apply(self, arguments); };
    for(var key in self) {
        if (self.hasOwnProperty(key)) {
            temp[key] = self[key];
        }
    }
    return temp;
};

function clone(obj){
    if (!(obj instanceof Object)) return obj
    const new_obj = new obj.constructor()
    for (const k in obj){
        new_obj[k] = null
    }
    for (const k in obj){
        const v = obj[k]
        if (v instanceof Function) new_obj[k] = cloneFun(v)
        else new_obj[k] = clone(v)
    }
    return new_obj
}


const baseModify = new SltThunk( 
    () => new SltFunc(
        (struct) => new SltFunc(
            (str) => new SltFunc(
                (fun) => {
                    const a = clone(struct())
                    console.log(a);
                    if (!(a instanceof SltStruct)) error("first argument TypeError", "Must be a struct rather than being " + a.type_, a)
                    if (str().value in a.value) {
                        const val = a.value[str().value]
                        console.log(val());
                        a.value[str().value] = new SltThunk(() => fun()(val))
                        return a
                    }
                    error("first argument KeyError",  "Key " + str().toString() + " not found in " + struct().toString(), a)
                },
                [1, 1, "modify.slt"]
            ),
            [1, 1, "modify.slt"]
        ), 
        [1, 1, "modify.js"]
    )
)

export {baseModify}